proposals/accepted/50-ide-test-integration.md

```diff
@@ -0,0 +1,219 @@
+"Re-run failed tests" as well as others. For users of mainstream
+languages, such features are no longer features, but instead
+expectations from a mature language.
+
+Haskell has a mature and well-maintained testing ecosystem, including
+the libraries: `tasty`, `HUnit`, and `hspec`.  While these libraries
+themselves are mature, there is currently minimal
+integration between Haskell's testing libraries and HLS, as well as
+the primary IDEs used for Haskell (VSC;Emacs;Neovim).
+
+Attempts at integration between testing frameworks have focused on
+regex searches, with support for hspec:
+ 1. [neo-test](https://github.com/mrcjkb/neotest-haskell)
+ 2. [vim-test](https://github.com/vim-test/vim-test)
+
+While these are valuable contributions to the ecosystem, their choice
+of regex for the collection of the user's test tree, and support for only
+specific frameworks are symptomatic of a lack of ecosystem coordination.
+
+## Motivation
+
+With the implementation of this proposal, the Haskell community will
+have gained a valuable IDE tool which will improve the user experience
+of Haskellers as well as make Haskellers more productive.
+
+Support for IDE integration of testing libraries will provide the following
+benefits:
+ 1. Ability for IDEs with Haskell support to provide inline hints for "Run
+ Tests" style commands.
+ 2. Ability for IDEs with Haskell support and support for a test explorer
+ view to display the user's test tree and run individual tests.
+ 3. A unified framework for testing libraries to provide integration with
+ IDEs. When a new testing framework is developed, they will be able to
+ provide the results in a Haskell datatype via serialization, which
+ avoids framework-specific test tree discovery logic.
+ 4. Potential to share test result rendering logic between different testing
+ frameworks, whereas currently each testing framework has their own test result
+ display logic.
+
+## Goals
+
+1. Create a new library, `hs-test`, with common datatypes for testing framework
+ test trees, and test results.
+2. Generate test trees and test results that can be serialized into JSON from
+Haskell's major testing frameworks, starting with `tasty`.
+3. Integrate test tree output output with VSC's Haskell extension, via VSC's [Testing API](https://code.visualstudio.com/api/extension-guides/testing).
+4. Stretch goal: support IDEs other than VSC.
+5. Stretch goal: support test results caching, and features such as "Rerun only failed tests".
+
+## Technical Proposal
+
+The `hs-test` library will contain the following data declarations:
+
+```haskell
+-- | Datatype representing all the tests that exist in a test
+-- suite that could be run.
+--
+-- `SrcLocs` are used to indicate where inline hints should be displayed
+-- in an IDE.
+data TestTree =
+  TestGroup {name :: Text, subTrees :: [TestTree], location :: SrcLoc}
+  | TestCase {name :: Text, location :: SrcLoc}
+
+data RunTestsRequest =
+  RunTestsRequest {path :: NonEmpty Text}
```

Comment:
Nothing in Tasty guarantees that the names are unique, and if I have my Tasty test tree factored out into a variety of top-level definitions, then I could very easily end up accidentally reusing a name. I would then be surprised if the wrong test is run when I click on it, and it might make me miss a bug.

Is there another way to represent the path?

For instance, what about a redundant encoding of name-position pairs?

Reply:
Working out what I imagine would happen here:
- The path gets turned into a tasty pattern
- The pattern matches multiple tests
- Multiple tests get run

Seems somewhat bad but maybe acceptable?